querybook/server/lib/table_upload/exporter/s3_exporter.py

```diff
@@ -43,15 +53,56 @@ def _upload_to_s3(self) -> None:
         """
         raise NotImplementedError()
 
-    def destination_s3_path(self) -> str:
-        return self.destination_s3_folder() + "/" + "0000"
+    @with_session
+    def destination_s3_path(self, session=None) -> str:
+        return self.destination_s3_folder(session=session) + "/" + "0000"
 
-    def destination_s3_folder(self) -> str:
-        schema_name, table_name = self._fq_table_name
-        object_key = re.sub(
-            S3_OBJECT_KEY_NOT_ALLOWED_CHAR, "", f"{schema_name}/{table_name}"
-        )
-        return QuerybookSettings.TABLE_UPLOAD_S3_PATH + object_key
+    @with_session
+    def destination_s3_root(self, session=None) -> str:
+        """Generate the bucket name + prefix before
+           the table specific folder
+
+        Returns:
+            str: s3 path consisting bucket + prefix + schema name
+        """
+        if "s3_path" in self._exporter_config:
+            schema_name, _ = self._fq_table_name
+            s3_path: str = self._exporter_config["s3_path"]
+            if not s3_path.endswith("/"):
+                s3_path += "/"
+            return s3_path + schema_name + "/"
+
+        if self._exporter_config.get("use_schema_location", False):
+            # Defer import since this is only needed for this option
+            from lib.metastore.loaders.hive_metastore_loader import HMSMetastoreLoader
+
+            query_engine = get_query_engine_by_id(self._engine_id, session=session)
+            metastore: HMSMetastoreLoader = get_metastore_loader(
+                query_engine.metastore_id, session=session
+            )
+            if metastore is None or not isinstance(metastore, HMSMetastoreLoader):
+                raise Exception("Invalid metastore to use use_schema_location option")
+            schema_location_uri = metastore.hmc.get_database(
+                self._table_config["schema_name"]
+            ).locationUri
+
+            # if the url is s3a or s3n, replace with s3
+            sanitized_schema_location_uri = re.sub(
+                r"^s3[a-z]:", "s3:", schema_location_uri
+            )
+
+            if not sanitized_schema_location_uri.endswith("/"):
+                sanitized_schema_location_uri += "/"
```

Comment:
can we have a separate util function to normalize a s3 path?

Also would it be possible that the s3_path from `_export_config` contains `s3a` or `s3n`?

Reply:
sounds good, added a shared function

as for the second point, no, because it is specified in the plugin, see documentation:
```
-   s3_path (str): if supplied, will use it as the root path for upload. Must be the full s3 path like s3://bucket/key, the trailing / is optional.
```

Reply:
for this share function, I meant to have it also include the `s3a`, `s3n` -> `s3` replacement.